server-spi/src/main/java/org/keycloak/storage/StorageId.java

```diff
@@ -21,35 +21,52 @@
 import org.keycloak.models.UserModel;
 
 import java.io.Serializable;
+import java.util.Objects;
 
 /**
  * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
  * @version $Revision: 1 $
  */
 public class StorageId implements Serializable {
-    private String id;
-    private String providerId;
-    private String externalId;
+    private final String providerId;
+    private final String externalId;
 
 
     public StorageId(String id) {
-        this.id = id;
         if (!id.startsWith("f:")) {
+            providerId = null;
             externalId = id;
-            return;
+        } else {
+            int providerIndex = id.indexOf(':', 2);
+            providerId = id.substring(2, providerIndex);
+            externalId = id.substring(providerIndex + 1);
         }
-        int providerIndex = id.indexOf(':', 2);
-        providerId = id.substring(2, providerIndex);
-        externalId = id.substring(providerIndex + 1);
-
     }
 
     public StorageId(String providerId, String externalId) {
-        this.id = "f:" + providerId + ":" + externalId;
+        if (providerId != null && providerId.contains(":")) {
```

Comment:
Question: If we have userStorage provider with providerId like "my:cool:provider" in current Keycloak, does it work or is it already broken due the colon in the providerId?

I just wonder if this is breaking change (In that case we may probably document this in the migration guide that people should remove colon from their providerId). But if providerId with colon is already broken in current Keycloak, we probably don't need to care about this IMO.

Reply:
It is in a way already broken, but since the provider ID is a component ID which in turn is a UUID, I believe we're still good

Reply:
Yes, true. I am now checking that it is UUID. I was mistakenly assuming that it is providerId as returned by `UserStorageProviderFactory.getId()` :-) Sorry for the confusion.